notebooks/Cross Validation Primer.ipynb

```diff
@@ -0,0 +1,432 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#this the same code as above\n",
+    "#except see how x_train is chosen differently\n",
+    "\n",
+    "#set the seed you get the same data each time you run\n",
+    "np.random.seed(0)\n",
+    "\n",
+    "#define a model to generate fake data\n",
+    "def data_model(x, sigma=0.1):\n",
+    "    base_model = np.log(x)\n",
+    "    noise = sigma * np.random.randn(len(x))\n",
+    "    return base_model + noise\n",
+    "\n",
+    "#Generate some fake data and keep a subset of them\n",
+    "#for training\n",
+    "#***Try different values for N***\n",
+    "N_split = 80\n",
+    "x_plot = np.linspace(1,100,num=100)\n",
+    "x = np.copy(x_plot)\n",
+    "\n",
+    "rng = np.random.RandomState(0)\n",
+    "rng.shuffle(x)\n",
+    "x_train = x[:N_split]\n",
+    "y_train = data_model(x_train)\n",
+    "\n",
+    "#also make validation data\n",
+    "x_val = x[N_split:]\n",
+    "y_val = data_model(x_val)\n",
+    "# create matrix versions of these arrays\n",
+    "X_train = x_train[:, np.newaxis]\n",
+    "X_val = x_val[:, np.newaxis]\n",
+    "X_plot = x_plot[:, np.newaxis]\n",
+    "\n",
+    "colors = ['teal', 'yellowgreen', 'gold', 'red']\n",
+    "lw = 0\n",
+    "plt.figure(figsize = (8,8))\n",
+    "plt.scatter(x_train, y_train, color='navy', s=30, marker='o', label=\"training points\", linewidth=lw)\n",
+    "plt.xlabel('x')\n",
+    "plt.ylabel('y')\n",
+    "\n",
+    "plt.scatter(x_val, y_val, marker='o', color='cornflowerblue', alpha=0.5,linewidth=lw, label=\"validation data\")\n",
+    "\n",
+    "plt.ylim([0,6])\n",
+    "lw =2\n",
+    "\n",
+    "models = [\n",
+    "    make_pipeline(PolynomialFeatures(2), Ridge()), # Overly Simple\n",
+    "    make_pipeline(PolynomialFeatures(4), Ridge()), # Just Right\n",
+    "    DecisionTreeRegressor(max_depth=80)  # Overly Complex\n",
+    "]\n",
+    "\n",
+    "labels = ['overly simple', 'just right', 'overly complex']\n",
+    "#for count, degree in enumerate([2, 5, 12]):\n",
+    "for count, model in enumerate(models):\n",
+    "    #fit to training data\n",
+    "    model.fit(X_train, y_train)\n",
+    "    y_train_fit = model.predict(X_train)\n",
+    "    nscore = skl.metrics.r2_score(y_train, y_train_fit)\n",
+    "    #apply to validation data\n",
+    "    y_predicted = model.predict(X_val)\n",
+    "    #get R^2 score\n",
+    "    score = skl.metrics.r2_score(y_val, y_predicted)\n",
+    "    #apply again, just to show plot\n",
+    "    y_plot = model.predict(X_plot)\n",
+    "    plt.plot(x_plot, y_plot, color=colors[count], linewidth=lw, alpha=0.5,\n",
+    "             label=labels[count] + \" with CV score %0.02f \\n   naive score %0.02f\" %(score, nscore))\n",
+    "\n",
+    "plt.legend(loc='lower right')\n",
```

Comment:
Add a `;` to the end of this line to get rid of the output
`<matplotlib.legend.Legend at XXXXXXXXX>`